webpack/webpack-embedded-dependencies-plugin/src/index.ts

```diff
@@ -0,0 +1,242 @@
+            if (packages.length === 0) {
+              compilation.warnings.push(
+                new WebpackError(
+                  `[embedded-dependencies-webpack-plugin]: No third party dependencies were found. Skipping license file generation.`
+                )
+              );
+              callback();
+              return;
+            }
+            // We should try catch here because generator function can be output from user config
+            try {
+              if (this.generateLicenseFileFunction) {
+                compilation.emitAsset(
+                  this.generatedLicenseFilename,
+                  new sources.RawSource(this.generateLicenseFileFunction(packages))
+                );
+              } else {
+                compilation.emitAsset(
+                  this.generatedLicenseFilename,
+                  new sources.RawSource(this._defaultLicenseFileGenerator(packages))
+                );
+              }
+            } catch (e) {
+              compilation.errors.push(
+                new WebpackError(
+                  `[embedded-dependencies-webpack-plugin]: Failed to generate license file: ${e}`
+                )
+              );
+            }
+          }
+
+          callback();
+        }
+      );
+    });
+  }
+
+  private async _getLicenseFilePath(modulePath: string, compiler: Compiler): Promise<string | undefined> {
+    type InputFileSystemReadDirResults = Parameters<
+      Parameters<typeof compiler.inputFileSystem.readdir>[1]
+    >[1];
+
+    const files: InputFileSystemReadDirResults = await LegacyAdapters.convertCallbackToPromise(
+      compiler.inputFileSystem.readdir,
+      modulePath
+    );
+
+    return files
+      ?.map((file) => file.toString())
+      .filter((file) => LICENSE_FILES_REGEXP.test(file))
+      .map((file) => path.join(modulePath, file))[0]; // Grabbing the first license file if multiple are found
+  }
+
+  /**
+   * Given a module path, try to parse the module's copyright attribution.
+   */
+  private async _parseCopyright(modulePath: string, compiler: Compiler): Promise<string | undefined> {
+    const licenseFile: string | undefined = await this._getLicenseFilePath(modulePath, compiler);
+
+    if (licenseFile) {
+      const license: string = await FileSystem.readFileAsync(licenseFile);
+      const match: RegExpMatchArray | null = license.match(COPYRIGHT_REGEX);
+
+      if (match) {
+        return match[0];
+      }
+    }
+
+    return undefined;
+  }
+
+  private _defaultLicenseFileGenerator(packages: IPackageData[]): string {
+    const licenseFileStrings: string[] = [];
+
+    const licenseTemplateForPackage = (pkg: IPackageData, licenseContent: string): string => {
+      return `
+        <hr />
+        ${pkg.name} - ${pkg.version}
+
+        ${licenseContent}
+      `;
+    };
+
+    for (const pkg of packages) {
+      if (pkg.licensePath) {
+        const licenseContent: string | undefined = FileSystem.readFile(pkg.licensePath);
```

Comment:
`_parseCopyright` will have already read this file. Can we read it in advance and include the raw content in the object being passed to the formatter?